File: Data_Structures/Assignment11/dijkstra.py
# dijkstra.py CS2028 2016 Cheng
# Dijkstra's Algorithm with heap for shortest path on a map

import turtle, math
import heap4

class Vertex:
	def __init__(self,vertexId,x,y,label):
		self.vertexId = vertexId
		self.x = x
		self.y = y
		self.label = label
		self.adjacentTo = []
		self.heapPosition = self.vertexId
		self.cost = 1000
		self.path = -1
        
class Edge:
	def __init__(self,v1,v2,weight,label):
		self.v1 = v1
		self.v2 = v2
		self.weight = weight
		self.label = label

vertexList = []
edgeList = []

def readGraph():
	file = open("roadGraph.txt", "r")
	N,L = file.readline().strip().split("\t")
	for i in range(int(N)):
		label,y,x = file.readline().strip().split("\t")
		vertexList.append(Vertex(i,float(x),float(y),label))
	for line in file:
		v1,v2,miles,road = line.strip().split("\t")
		edgeList.append(Edge(int(v1), int(v2), int(miles), road))
	file.close()

def addNeighbors():
	for edge in edgeList:
		vertexList[edge.v1].adjacentTo.append(edge)
		vertexList[edge.v2].adjacentTo.append(edge)


def showGraph():
	t = turtle.Turtle()
	screen = t.getscreen()
	minx = 1000.0
	maxx = -1000.0
	miny = 1000.0
	maxy = -1000.0
	for vertex in vertexList:
		if vertex.x > maxx:
			maxx = vertex.x
		if vertex.x < minx:
			minx = vertex.x
		if vertex.y > maxy:
			maxy = vertex.y
		if vertex.y < miny:
			miny = vertex.y
	ax = 560.0 /(maxx - minx)
	bx = -280.0 - ax * minx
	ay = 560.0 /(maxy - miny)
	by = -280.0 - ay * miny
	for vertex in vertexList:
		t.penup()
		t.goto(vertex.x * ax + bx,vertex.y * ay + by)
		t.pendown()
		if vertex.heapPosition == -2:
			t.color('red')
		t.write(vertex.label,align="center",font=("Arial",7,"bold"))
		if vertex.heapPosition == -2:
			t.color('black')
	for edge in edgeList:
		t.penup()
		x1 = vertexList[edge.v1].x * ax + bx
		y1 = vertexList[edge.v1].y * ay + by
		x2 = vertexList[edge.v2].x * ax + bx
		y2 = vertexList[edge.v2].y * ay + by
		t.goto(x1,y1)
		t.pendown()
		t.goto(x2,y2)		
	t.ht()
	screen.exitonclick()
	print("Program Execution Completed.")

def binarySearch(label):
	lo = 0
	hi = len(vertexList) - 1
	while lo <= hi:
		mid = (lo + hi) // 2
		if label == vertexList[mid].label:
			return mid
		if label < vertexList[mid].label:
			hi = mid - 1
		else:
			lo = mid + 1
	return -1

def getLocation(locationType):
	while True:
		location = input('Enter a ' + locationType + ': ')
		if len(location) < 1:
			return -1
		index = binarySearch(location)
		if index >= 0:
			return index
		else:
			print("no such location")
	return -1

def dijkstra(source, destination):
	vertexList[source].cost = 0
	heap = heap4.Heap(vertexList)
	heap.siftUpFrom(source)
	while heap.size > 0:
		top = heap.removeTop()
		if top == destination:
			printPath(top)
			return
		current = vertexList[top]
		for edge in current.adjacentTo:
			if edge.v1 == top:
				adjacent = vertexList[edge.v2]
			else:
				adjacent = vertexList[edge.v1]
			if adjacent.heapPosition >= 0:
				if current.cost + edge.weight < adjacent.cost:
					adjacent.cost = current.cost + edge.weight
					adjacent.path = top
					heap.siftUpFrom(adjacent.heapPosition)

def printPath(v):
	if vertexList[v].path >= 0:
		printPath(vertexList[v].path)
	print(vertexList[v].label)
	vertexList[v].heapPosition = -2

def main():
	readGraph()
	addNeighbors()
	source = getLocation('source')
	if source < 0:
		return
	destination = getLocation('destination')
	if destination < 0:
		return
	dijkstra(source, destination)
	showGraph()

if __name__ == "__main__":
	main()

File: Data_Structures/Assignment11/dijkstra.cpp
// dijkstra.cpp CS2028 2016 cheng
// Dijkstra's algorithm with heap as priority queue
// To be used with roadGraph.txt or graphs with similar format

#include <iostream>
#include <fstream>
#include "heap2.h"
using namespace std;

int N, L;
Vertex *vertexList;
Edge *edgeList;

void readGraph(){
    ifstream inputfile;
    inputfile.open("roadGraph.txt");
    if (!inputfile) return;
    inputfile >> N >> L;
    vertexList = new Vertex[N];
    edgeList = new Edge[L];
    string x; string y;
    for (int i = 0; i < N; i++){
        inputfile >> vertexList[i].label >> x >> y;
        vertexList[i].heapPosition = i;
        vertexList[i].cost = 1000;
        vertexList[i].path = -1;
    }
    int u1; int u2; int w; string label;
    for (int i = 0; i < L; i++){
        inputfile >> u1 >> u2 >> w >> label;
        edgeList[i].v1 = u1;
        edgeList[i].v2 = u2;
        edgeList[i].weight = w;
    }
    inputfile.close();
}

void addNeighbors(){
    for (int i = 0; i < L; i++){
        vertexList[edgeList[i].v1].adjacentTo.push_back(edgeList + i);
        vertexList[edgeList[i].v2].adjacentTo.push_back(edgeList + i);
    }
}

void printGraph(){  // How to go through a vecter and 2 ways to use pointers to pointers
    for (int i = 0; i < N; i++){
        cout << vertexList[i].label << endl;
        for (vector<Edge*>::iterator iter = vertexList[i].adjacentTo.begin();
            iter < vertexList[i].adjacentTo.end(); iter++)
            cout << "\t" << vertexList[(**iter).v1].label << "\t"
                << vertexList[(*iter)->v2].label << endl;
    }
}

void printPath(int v){
    if (vertexList[v].path >= 0) printPath(vertexList[v].path);
    cout << vertexList[v].label << endl;
}

void dijkstra(int source, int destination){
    vertexList[source].cost = 0;
    Heap heap(vertexList, N);
    heap.siftUpFrom(source);
    while (heap.size > 0){
        int top = heap.removeTop();
        if (top == destination){
            printPath(top);
            return;
        }
        for (vector<Edge*>::iterator iter = vertexList[top].adjacentTo.begin();
            iter < vertexList[top].adjacentTo.end(); iter++){
            int adjacent = (**iter).v1 == top ? (**iter).v2 : (**iter).v1;
            if (vertexList[adjacent].heapPosition >= 0)
                if (vertexList[top].cost + (**iter).weight < vertexList[adjacent].cost){
                    // Your code here to update vertexList[adjacent].cost
                    // and then sift it up from its current heap position
                    vertexList[adjacent].cost = vertexList[top].cost + (**iter).weight;
                    vertexList[adjacent].path = top;
                    heap.siftUpFrom(vertexList[adjacent].heapPosition);
                }
        }
    }
}

int main(){
    readGraph();
    addNeighbors();
    dijkstra(7, 13);
}



